NEW Add backing off to failed factoryCreate for node-pool

dhensby opened the pull request:
What this does:
This is a potential fix to #805 

~~It's still work in progress as there are some `UnhandledPromiseRejectionWarning`s, which I'm trying to get to the bottom of.~~

Thoughts on this resolution would be appreciated

---

NB: It seems that there is a problem with generic-pool where it will continually attempt to create resources no matter what and there's not a lot we can do to resolve that.

Not only that, it will continue to attempt to create resources after `drain` is called (attempting to close the pool) - I think because they are already queued up.

dhensby commented on lib/base.js:
```diff
@@ -250,9 +250,29 @@ class ConnectionPool extends EventEmitter {
+            if (!this.connecting && !this.connected) {
+              throw new ConnectionError('Cannot create connections on closed pool')
+            }
+            if (Date.now() > this.pool._config.acquireTimeoutMillis + start) {
```
We have to implement our own timeout, which is a pain

dhensby commented:
OMG - this might go green!

dhensby commented:
Although we are in beta for v5, I think this is probably sufficiently important enough a bug to patch in the beta phase. 

Note that this fixes an issue where `generic-pool` just hammers the DB server attempting to connect to the DB and therefore blocking the event loop. It does *not* prevent the pool from trying to create connections after the acquire call fails to create a resource, so it still could theoretically attempt to create connections forever.

dhensby commented on lib/base.js:
```diff
@@ -251,20 +251,24 @@ class ConnectionPool extends EventEmitter {
+                throw new ConnectionError('Connection is closed.', 'ECONNCLOSED')
+              }
+              return new Promise((resolve) => {
+                console.log('backing off', backoff)
```
oh gosh, yes, that shouldn't be there - I'll look at putting some debug calls in

dhensby commented:
@clshortfuse this is your chance to give some feedback on this fix

clshortfuse commented:
I'll have to check it on my AWS server and Mac OS X environment a bit later today. Just glancing at the code, it would be nice to have an immediate retry if retryCount is 0. 2**0 would set it to 1ms which is probably close enough.

I'm a little paranoid about infinitely increasing the exponents of 2, since I'm not sure how all environments would handle 2**64 * 100. Something tells me one environment could return -1 or perhaps throw an error. Once you reach the MAX_TIMEOUT (30000ms), there's no reason to keep increasing calculating the increased backoff.

dhensby commented:
> I'm a little paranoid about infinitely increasing the exponents of 2, since I'm not sure how all environments would handle 2**64 * 100. Something tells me one environment could return -1 or perhaps throw an error. Once you reach the MAX_TIMEOUT (30000ms), there's no reason to keep increasing calculating the increased backoff.

Yes, I had this thought too - I'll look into it

dhensby commented:
> it would be nice to have an immediate retry if retryCount is 0

I'm not sure why an immediate retry would be desirable on failure; if the server is down we want to give it some time to recover and 200ms seems fair enough

dhensby commented:
@clshortfuse OK - I've addressed your feedback so far and I think this is as ready as it's going to be

@willmorgan care to review / approve / merge? I think I'll then tag v5 as stable as it's been long enough...

Do we need some docs around this retry strategy?

willmorgan commented on lib/base.js:
```diff
@@ -250,9 +250,46 @@ class ConnectionPool extends EventEmitter {
+              }).then(run.bind(this))
+            })
+          }
+          return run.apply(this)
```
this can be call, since no arguments are being passed

willmorgan commented on lib/base.js:
```diff
@@ -250,9 +250,46 @@ class ConnectionPool extends EventEmitter {
+          let backoff
+          let retryCount = 0
+          const self = this
+          const run = function () {
```
function run rather than const run?

willmorgan commented on lib/base.js:
```diff
@@ -250,9 +250,46 @@ class ConnectionPool extends EventEmitter {
+                }, backoff)
+                timer.unref()
+                debug('pool(%d): failed to create connection resource(%d), retrying with backoff of %dms', IDS.get(self), createId, backoff)
+              }).then(run.bind(this))
```
I don't think you need to `.bind` here because you're already referencing `self` (which could be renamed to something better)

dhensby replied:
I'm mixing use of `self` and `this` which is a bit gross - `self` is mostly for the Promise function, though...

dhensby replied:
`self` is a bit of a bad name, but the only sane name is `connectionPool` but that's pretty ambiguous, so I think I'm going to leave it as is

willmorgan replied:
In any case you need not re-bind.

dhensby replied:
removing the `bind` breaks the behaviour

willmorgan commented on lib/base.js:
```diff
@@ -250,9 +250,46 @@ class ConnectionPool extends EventEmitter {
 
+        let createAttempt = 0
+
+        function create () {
```
I wonder if this function can be brought out of the closure here to save on redefining it every time connect is called?

dhensby replied:
Connect isn't really called regularly and it currently helps keep scope of the vars a bit more manageable (IMO)

willmorgan replied:
I do think that these functions should be extracted and parameterised to avoid nesting too deeply, and to allow easy extension of `ConnectionPool` in userland code. It doesn't look like there would be much to parameterise.

dhensby replied:
OK - I've looked at this and it just makes scoping a bit too hard. This feature is also going to be removed in v6 because tarn has a retry strategy that can be configured.

As such we don't want to make this available on the `ConnectionPool` for userland changes and the scoping of the retry count becomes too difficult if we move the function around...

dhensby replied:
OK - I've managed to pull it out ;)

willmorgan commented on lib/base.js:
```diff
@@ -250,9 +250,46 @@ class ConnectionPool extends EventEmitter {
+
+        function create () {
+          const createId = createAttempt++
+          let timeout = this.config.pool ? this.config.pool.acquireTimeoutMillis : undefined
```
We're going to need some docs on how to specify this.

I don't think the timeout should be infinite by default as none of this useful work would actually get executed. Maybe 32s (4 retry attempts) is sensible?

dhensby replied:
I don't think we should be changing the default values. It's up to devs to define their acquireTimeouts and some people seem to want a never-ending acquire timeout...

willmorgan replied:
Let's revisit this once in the wild.
